### PR TITLE
Helper methods for device attestation  PR

### PR DESCRIPTION
**Pull request by mleisner**
#### Problem
Provide data access methods for device attestation PR (#8435) 

#### Change overview
- Construct Attestation Elements TLV
- Deconstruct Attestation Elements TLV
- Sign and Validate Attestation Info
- Extract Authority Key Id from x509 Certificate
- Extract Vendor ID from Attestation x509 Certificate
- Added new unit tests to support testing of new methods
- Added Generic Implementation of methods to retrieve DAC/PAI/PAA Certificates
- Added Generic Implementation of methods to retrieve Certification Declaration
- Added Generic Implementation of methods to retrieve Attestation Signature

#### Testing
Matter unit tests

### CODE REVIEW

File: src/crypto/CHIPCryptoPAL.cpp [hunk 1]
```diff
@@ -648,5 +649,185 @@ CHIP_ERROR GenerateCompressedFabricId(const Crypto::P256PublicKey & root_public_
...
+        ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(6), vendorReserved6));
+    }
+    if (!vendorReserved7.empty())
+    {
+        ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(7), vendorReserved7));
+    }
+    if (!vendorReserved8.empty())
+    {
+        ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(8), vendorReserved8));
+    }
```
Comment (tcarmelveilleux) [inline src/crypto/CHIPCryptoPAL.cpp:720]: The scheme for vendor-reserved used here is not spec compliant. Recent changes were made to normalize it, see https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/3598/files

File: src/crypto/CHIPCryptoPAL.cpp [hunk 2]
```diff
@@ -648,5 +649,185 @@ CHIP_ERROR GenerateCompressedFabricId(const Crypto::P256PublicKey & root_public_
...
+    TLV::TLVReader tlvReader;
+    TLV::TLVType containerType = TLV::kTLVType_Structure;
+
+    tlvReader.Init(attestationElements.data(), static_cast<uint32_t>(attestationElements.size()));
+    ReturnErrorOnFailure(tlvReader.Next(containerType, TLV::AnonymousTag));
+    ReturnErrorOnFailure(tlvReader.EnterContainer(containerType));
+
+    memset(argExists, 0, sizeof argExists);
+
+    for (size_t i = 0; i < sizeof(element_array) / sizeof(*element_array); ++i)
```
Comment (tcarmelveilleux) [inline src/crypto/CHIPCryptoPAL.cpp:750]: There is `ArraySize` in `src/lib/support/CodeUtils.h`

File: src/crypto/CHIPCryptoPAL.cpp [hunk 3]
```diff
@@ -648,5 +649,185 @@ CHIP_ERROR GenerateCompressedFabricId(const Crypto::P256PublicKey & root_public_
...
+    CHIP_ERROR TLVUnpackError   = CHIP_NO_ERROR;
+    bool argExists[9];
+    TLV::TLVReader tlvReader;
+    TLV::TLVType containerType = TLV::kTLVType_Structure;
+
+    tlvReader.Init(attestationElements.data(), static_cast<uint32_t>(attestationElements.size()));
+    ReturnErrorOnFailure(tlvReader.Next(containerType, TLV::AnonymousTag));
+    ReturnErrorOnFailure(tlvReader.EnterContainer(containerType));
+
+    memset(argExists, 0, sizeof argExists);
```
Comment (tcarmelveilleux) [inline src/crypto/CHIPCryptoPAL.cpp:748]:
```suggestion
    memset(argExists, 0, sizeof(argExists));
```
for consistency

File: src/crypto/CHIPCryptoPAL.cpp [hunk 4]
```diff
@@ -648,5 +649,185 @@ CHIP_ERROR GenerateCompressedFabricId(const Crypto::P256PublicKey & root_public_
...
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DeconstructAttestationElements(const ByteSpan & attestationElements, ByteSpan & certificationDeclaration,
+                                          ByteSpan & attestationNonce, uint32_t & timestamp, ByteSpan & firmwareInfo,
+                                          ByteSpan & vendorReserved5, ByteSpan & vendorReserved6, ByteSpan & vendorReserved7,
+                                          ByteSpan & vendorReserved8)
+{
+    ByteSpan * element_array[] = { &certificationDeclaration, &attestationNonce, nullptr,          &firmwareInfo,
```
Comment (tcarmelveilleux) [inline src/crypto/CHIPCryptoPAL.cpp:733]: This should be an array of references

File: src/crypto/CHIPCryptoPAL.cpp [hunk 5]
```diff
@@ -648,5 +649,185 @@ CHIP_ERROR GenerateCompressedFabricId(const Crypto::P256PublicKey & root_public_
...
+    ReturnErrorOnFailure(tlvReader.Next(containerType, TLV::AnonymousTag));
+    ReturnErrorOnFailure(tlvReader.EnterContainer(containerType));
+
+    memset(argExists, 0, sizeof argExists);
+
+    for (size_t i = 0; i < sizeof(element_array) / sizeof(*element_array); ++i)
+    {
+        if (element_array[i] != nullptr)
+        {
+            *element_array[i] = ByteSpan();
```
Comment (tcarmelveilleux) [inline src/crypto/CHIPCryptoPAL.cpp:754]: Since this is not an array of references, this will not overwrite the ouput

File: src/crypto/CHIPCryptoPAL.cpp [hunk 6]
```diff
@@ -648,5 +649,185 @@ CHIP_ERROR GenerateCompressedFabricId(const Crypto::P256PublicKey & root_public_
...
+    }
+
+    while ((TLVError = tlvReader.Next()) == CHIP_NO_ERROR)
+    {
+        // Since call to aDataTlv.Next() is CHIP_NO_ERROR, the read head always points to an element.
+        // Skip this element if it is not a ContextTag, not consider it as an error if other values are valid.
+        if (!TLV::IsContextTag(tlvReader.GetTag()))
+        {
+            continue;
+        }
```
Comment (tcarmelveilleux) [inline src/crypto/CHIPCryptoPAL.cpp:765]: Vendor-specific tags do not fit with this method of access since they are fully qualified tags.

File: src/crypto/CHIPCryptoPAL.cpp [hunk 7]
```diff
@@ -648,5 +649,185 @@ CHIP_ERROR GenerateCompressedFabricId(const Crypto::P256PublicKey & root_public_
...
+        currentDecodeTagId = TLV::TagNumFromTag(tlvReader.GetTag());
+        if (currentDecodeTagId < 9 && currentDecodeTagId > 0)
```
Comment (tcarmelveilleux) [inline src/crypto/CHIPCryptoPAL.cpp:767]: This is confusing logic in that order.

File: src/crypto/CHIPCryptoPAL.cpp [hunk 8]
```diff
@@ -648,5 +649,185 @@ CHIP_ERROR GenerateCompressedFabricId(const Crypto::P256PublicKey & root_public_
...
+        case 7:
+        case 8: {
+            const uint8_t * data = nullptr;
+            TLVUnpackError       = tlvReader.GetDataPtr(data);
+            if (element_array[currentDecodeTagId - 1] != nullptr)
+            {
+                *element_array[currentDecodeTagId - 1] = ByteSpan(data, tlvReader.GetLength());
+            }
+        }
+        break;
```
Comment (tcarmelveilleux) [inline src/crypto/CHIPCryptoPAL.cpp:797]: Move break inside the switch scoped block

File: src/crypto/CHIPCryptoPAL.cpp [hunk 9]
```diff
@@ -648,5 +649,185 @@ CHIP_ERROR GenerateCompressedFabricId(const Crypto::P256PublicKey & root_public_
...
+        case 3: {
+            TLVUnpackError = tlvReader.Get(timestamp);
+        }
```
Comment (tcarmelveilleux) [inline src/crypto/CHIPCryptoPAL.cpp:800]: This is a fallthrough to the `default` break. Please add a break in case3

File: src/crypto/CHIPCryptoPAL.cpp [hunk 10]
```diff
@@ -648,5 +649,185 @@ CHIP_ERROR GenerateCompressedFabricId(const Crypto::P256PublicKey & root_public_
...
+        default:
+            // Unsupported tag, ignore it.
```
Comment (tcarmelveilleux) [inline src/crypto/CHIPCryptoPAL.cpp:802]: Only some unsupported tags are OK. They are the vendor-specific tags. This default should have an error depending on  the validity of the schema (e.g. if it's not a fully qualified tag, after all context-specific tags, then it's an error).

File: src/crypto/CHIPCryptoPAL.cpp [hunk 11]
```diff
@@ -648,5 +649,185 @@ CHIP_ERROR GenerateCompressedFabricId(const Crypto::P256PublicKey & root_public_
...
+        {
+            break;
+        }
+    }
+
+    if (CHIP_END_OF_TLV == TLVError)
+    {
+        // CHIP_END_OF_TLV means we have iterated all items in the structure, which is not a real error.
+        TLVError = CHIP_NO_ERROR;
+    }
```
Comment (tcarmelveilleux) [inline src/crypto/CHIPCryptoPAL.cpp:815]: This omits validating that all expected fields were read.

File: src/crypto/CHIPCryptoPAL.cpp [hunk 12]
```diff
@@ -648,5 +649,185 @@ CHIP_ERROR GenerateCompressedFabricId(const Crypto::P256PublicKey & root_public_
...
+        // Since call to aDataTlv.Next() is CHIP_NO_ERROR, the read head always points to an element.
+        // Skip this element if it is not a ContextTag, not consider it as an error if other values are valid.
+        if (!TLV::IsContextTag(tlvReader.GetTag()))
+        {
+            continue;
+        }
+        currentDecodeTagId = TLV::TagNumFromTag(tlvReader.GetTag());
+        if (currentDecodeTagId < 9 && currentDecodeTagId > 0)
+        {
+            if (argExists[currentDecodeTagId])
```
Comment (tcarmelveilleux) [inline src/crypto/CHIPCryptoPAL.cpp:769]: It's odd that the first item is skipped in here.

Overall, it's not obvious at all how schema is validated for correctness in this method

File: src/crypto/CHIPCryptoPAL.h [hunk 13]
```diff
@@ -286,6 +292,17 @@ class ECPKeypair
...
+    /**
+     * @brief A function to sign an attestation data blob using ECDSA
+     * @param attestationElements Span of the AttestationElements to be signed
+     * @param attestationChallenge Span of the AttestationChallenge to be signed
+     * @param out_signature Buffer that will hold the output signature. The signature consists of: 2 EC elements (r and s),
+     * in raw <r,s> point form (see SEC1).
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    virtual CHIP_ERROR ECDSA_sign_attestation_data(const ByteSpan & attestationElements, const ByteSpan & attestationChallenge,
+                                                   Sig & out_signature) = 0;
```
Comment (tcarmelveilleux) [inline src/crypto/CHIPCryptoPAL.h:304]: This does not belong in ECPKeyPair public API. Attestation signature is always P256 and this special casing with the attestation challenge should just be done by *using* a keypair, rather than being *part of* keypair.

File: src/crypto/CHIPCryptoPALOpenSSL.cpp [hunk 14]
```diff
@@ -1665,5 +1666,60 @@ CHIP_ERROR ExtractPubkeyFromX509Cert(const ByteSpan & certificate, Crypto::P256P
...
+    VerifyOrExit(x509certificate != nullptr, err = CHIP_ERROR_NO_MEMORY);
+
+    akidString = X509_get0_authority_key_id(x509certificate);
+
+    VerifyOrExit(akidString->length == static_cast<int>(akid.size()), err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+
+    for (int i = 0; i < akidString->length; ++i)
+    {
+        akid.data()[i] = akidString->data[i];
+    }
```
Comment (tcarmelveilleux) [inline src/crypto/CHIPCryptoPALOpenSSL.cpp:1687]: replace this manual copy loop with a `memcpy` to `akid.data()`

File: src/crypto/CHIPCryptoPALOpenSSL.cpp [hunk 15]
```diff
@@ -1665,5 +1666,60 @@ CHIP_ERROR ExtractPubkeyFromX509Cert(const ByteSpan & certificate, Crypto::P256P
...
+
+CHIP_ERROR ExtractVIDFromX509Cert(const ByteSpan & certificate, uint16_t & vid)
+{
+    CHIP_ERROR err                        = CHIP_NO_ERROR;
+    X509 * x509certificate                = nullptr;
+    const unsigned char * pCertificate    = certificate.data();
+    const unsigned char ** ppCertificate  = &pCertificate;
+    char * subj                           = nullptr;
+    char * vidSubj                        = nullptr;
+    static constexpr char vidNeedle[]     = "1.3.6.1.4.1.37244.2.1=";
```
Comment (tcarmelveilleux) [inline src/crypto/CHIPCryptoPALOpenSSL.cpp:1703]: Neither `static` argument need to be static

File: src/crypto/CHIPCryptoPALOpenSSL.cpp [hunk 16]
```diff
@@ -1665,5 +1666,60 @@ CHIP_ERROR ExtractPubkeyFromX509Cert(const ByteSpan & certificate, Crypto::P256P
...
+    x509certificate = d2i_X509(NULL, ppCertificate, static_cast<long>(certificate.size()));
+    VerifyOrExit(x509certificate != nullptr, err = CHIP_ERROR_NO_MEMORY);
+
+    subj = X509_NAME_oneline(X509_get_subject_name(x509certificate), NULL, 0);
+    VerifyOrExit(subj != nullptr, err = CHIP_ERROR_NO_MEMORY);
+
+    vidSubj = strstr(subj, vidNeedle);
+    VerifyOrExit(vidSubj != nullptr, err = CHIP_ERROR_WRONG_CERT_TYPE);
+
+    vid = static_cast<uint16_t>(strtol(vidSubj + vidNeedleSize, NULL, 16));
```
Comment (tcarmelveilleux) [inline src/crypto/CHIPCryptoPALOpenSSL.cpp:1715]: This needs to be `strtoul`. This also does not validate format. It's probably fine not to check for exact width, but we should check for error (e.g. not pass NULL for second arg). Overall, this likely will fail if the subject does not *end* with a VID, since the `strtoul` will see a subject separator.

Ex: `CN=Matter Test DAC 000C,1.3.6.1.4.1.37244.2.1=FFF1,1.3.6.1.4.1.37244.2.2=8001`

The method won't work there.

Also, docs for openssl state:
```
NOTES
The functions X509_NAME_oneline() and X509_NAME_print() are legacy functions which produce a non standard output form, they don't handle multi character fields and have various quirks and inconsistencies. Their use is strongly discouraged in new applications.
```

Please find an alternate method to extract VID

File: src/crypto/CHIPCryptoPALmbedTLS.cpp [hunk 17]
```diff
@@ -1207,5 +1207,15 @@ CHIP_ERROR ExtractPubkeyFromX509Cert(const ByteSpan & certificate, Crypto::P256P
...
 
+CHIP_ERROR ExtractAKIDFromX509Cert(const ByteSpan & certificate, MutableByteSpan & akid)
+{
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+
+CHIP_ERROR ExtractVIDFromX509Cert(const ByteSpan & certificate, uint16_t & vid)
+{
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
```
Comment (tcarmelveilleux) [inline src/crypto/CHIPCryptoPALmbedTLS.cpp:1218]: Please ensure arguments are ignored to avoid compile errors:

`(void)certificate`, etc

File: src/crypto/CHIPCryptoPAL.cpp [hunk 18]
```diff
@@ -648,5 +649,185 @@ CHIP_ERROR GenerateCompressedFabricId(const Crypto::P256PublicKey & root_public_
...
+    {
+        return TLVUnpackError;
+    }
+    if (validArgumentCount > 8)
+    {
+        return CHIP_ERROR_INVALID_TLV_ELEMENT;
+    }
+    return CHIP_NO_ERROR;
+}
+
```
[the remark this answers is not in the document]
Comment (emargolis) [reply src/crypto/CHIPCryptoPAL.cpp:831]: Suggest to move it to: src/credentials/DeviceAttestation.*

File: src/lib/core/tests/TestCHIPTLV.cpp [hunk 19]
```diff
@@ -2983,7 +2983,7 @@ void TestCHIPTLVReaderDup(nlTestSuite * inSuite)
 void TestCHIPTLVReaderErrorHandling(nlTestSuite * inSuite)
 {
     CHIP_ERROR err;
-    uint8_t buf[2048];
+    uint8_t buf[2048] = { 0 };
```
Comment (tcarmelveilleux) [inline src/lib/core/tests/TestCHIPTLV.cpp:2986]: Why this change?

File: src/include/platform/internal/GenericConfigurationManagerImpl.h [hunk 20]
```diff
@@ -77,6 +77,13 @@ class GenericConfigurationManagerImpl
     CHIP_ERROR _GetDeviceCertificate(uint8_t * buf, size_t bufSize, size_t & certLen);
     CHIP_ERROR _GetDeviceIntermediateCACerts(uint8_t * buf, size_t bufSize, size_t & certsLen);
     CHIP_ERROR _GetDevicePrivateKey(uint8_t * buf, size_t bufSize, size_t & keyLen);
+    CHIP_ERROR _GetDeviceDACCertificate(uint8_t * buf, size_t bufSize, size_t & dacLen);
+    CHIP_ERROR _GetDevicePAICertificate(uint8_t * buf, size_t bufSize, size_t & paiLen);
+    CHIP_ERROR _GetDevicePAACertificate(const uint8_t * skid, size_t skidLen, uint8_t * buf, size_t bufSize, size_t & paaLen);
+    CHIP_ERROR _GetDeviceCertificationDeclaration(uint8_t * buf, size_t bufSize, size_t & certDeclLen);
+    CHIP_ERROR _GetDeviceAttestationSignature(const uint8_t * attestationElements, size_t attestationElementsSize,
+                                              const uint8_t * attestationChallenge, size_t attestationChallengeSize, uint8_t * buf,
+                                              size_t bufSize, size_t & signatureLen);
```
Comment (tcarmelveilleux) [inline src/include/platform/internal/GenericConfigurationManagerImpl.h:86]: Having these default implementations of specific certificates here will make it harder to get to the real data per platform, and will also cost flash for the PAA on target devices. The API for accessing PAAs should not be baked here, as it is implementation-specific per commissioner. Lookup of PAA may not even be needed if attestation if verified through a cloud or out-of-SDK app logic. Suggest removing PAA getter from here.

File: src/include/platform/internal/GenericConfigurationManagerImpl.h [hunk 21]
```diff
@@ -77,6 +77,13 @@ class GenericConfigurationManagerImpl
     CHIP_ERROR _GetDeviceCertificate(uint8_t * buf, size_t bufSize, size_t & certLen);
     CHIP_ERROR _GetDeviceIntermediateCACerts(uint8_t * buf, size_t bufSize, size_t & certsLen);
     CHIP_ERROR _GetDevicePrivateKey(uint8_t * buf, size_t bufSize, size_t & keyLen);
+    CHIP_ERROR _GetDeviceDACCertificate(uint8_t * buf, size_t bufSize, size_t & dacLen);
+    CHIP_ERROR _GetDevicePAICertificate(uint8_t * buf, size_t bufSize, size_t & paiLen);
+    CHIP_ERROR _GetDevicePAACertificate(const uint8_t * skid, size_t skidLen, uint8_t * buf, size_t bufSize, size_t & paaLen);
+    CHIP_ERROR _GetDeviceCertificationDeclaration(uint8_t * buf, size_t bufSize, size_t & certDeclLen);
+    CHIP_ERROR _GetDeviceAttestationSignature(const uint8_t * attestationElements, size_t attestationElementsSize,
```
Comment (tcarmelveilleux) [inline src/include/platform/internal/GenericConfigurationManagerImpl.h:84]: The attestation primitive must be more generic and not rely on attestation elements since it's *also required for CSR*.

I have a suggested interface:
```
namespace chip {

class DeviceAttestationCredentials
{
public:
    DeviceAttestationCredentials() = default;
    virtual ~DeviceAttestationCredentials() = default;

    // Not copyable
    DeviceAttestationCredentials(const DeviceAttestationCredentials&) = delete;
    DeviceAttestationCredentials& operator=(const DeviceAttestationCredentials&) = delete

    virtual CHIP_ERROR GetCertificationDeclaration(MutableByteSpan & out_cd_buffer) = 0;
    virtual CHIP_ERROR GetFirmwareInformation(MutableByteSpan & out_firmware_info_buffer) = 0;
    virtual CHIP_ERROR GetDeviceAttestationCert(MutableByteSpan & out_dac_buffer) = 0;
    virtual CHIP_ERROR GetProductAttestationIntermediateCert(MutableByteSpan & out_pai_buffer) = 0;
    virtual CHIP_ERROR SignWithDeviceAttestationKey(const ByteSpan & payload, MutableByteSpan & out_signature_buffer) = 0;
};

}  // namespace chip

```

The above interface could be implemented per platform, and supports:

* Accessing the necessary fields for a device being attested, wtihout needing access to PAA
* Signing with the device attestation key without relying on having direct access to the key, and without relying on particular format

For the PAA certificates, I would argue that:

1. Device attestation signature + chain verification should be abstracted in a delegate that different platforms can implement.
2. While an example for chip-tool's commissioning code can be employed, it's important not to bake-in the assumption of direct access to PAA possible. It's already clear in spec that direct access to all PAAs is not mandatory at the commissioner itself (although it's feasible), due to the variety of device attestation policies.

For example: an ecosystem with full end-to-end attestation validation may require verifying the device attestation signature of the CSRResponse **at the CA** to ensure that the CA is not used as a signing oracle for operational credentials. Because of this, we should not bake a 1-size-fits all *validation* path with mandatory access to PAAs.

File: src/crypto/CHIPCryptoPALOpenSSL.cpp (hunk 16 above)
Comment (tcarmelveilleux) [reply src/crypto/CHIPCryptoPALOpenSSL.cpp:1715]: @emargolis Aren't there other methods already in tree to extract RDNs from x.509 subject fields?

File: src/include/platform/internal/GenericConfigurationManagerImpl.h (hunk 21 above)
Comment (emargolis) [reply src/include/platform/internal/GenericConfigurationManagerImpl.h:84]: @tcarmelveilleux  are you suggesting to define a separate class instead of using CRTP pattern as it is currently implemented in this file?

File: src/lib/core/tests/TestCHIPTLV.cpp (hunk 19 above)
Comment (emargolis) [reply src/lib/core/tests/TestCHIPTLV.cpp:2986]: There was a separate discussion that this buffer is not initialize and may actually include some valid TLV components.

File: src/crypto/CHIPCryptoPALOpenSSL.cpp [hunk 22]
```diff
@@ -1665,5 +1666,60 @@ CHIP_ERROR ExtractPubkeyFromX509Cert(const ByteSpan & certificate, Crypto::P256P
...
+    static constexpr char vidNeedle[]     = "1.3.6.1.4.1.37244.2.1=";
+    static constexpr size_t vidNeedleSize = sizeof(vidNeedle) - 1;
+
+    x509certificate = d2i_X509(NULL, ppCertificate, static_cast<long>(certificate.size()));
+    VerifyOrExit(x509certificate != nullptr, err = CHIP_ERROR_NO_MEMORY);
+
+    subj = X509_NAME_oneline(X509_get_subject_name(x509certificate), NULL, 0);
+    VerifyOrExit(subj != nullptr, err = CHIP_ERROR_NO_MEMORY);
+
+    vidSubj = strstr(subj, vidNeedle);
```
Comment (bzbarsky-apple) [inline src/crypto/CHIPCryptoPALOpenSSL.cpp:1712]: This seems extremely unsafe.  What if one of the values in the subject contains that "needle" string?

File: src/crypto/CHIPCryptoPALmbedTLS.cpp [hunk 23]
```diff
@@ -1207,5 +1207,15 @@ CHIP_ERROR ExtractPubkeyFromX509Cert(const ByteSpan & certificate, Crypto::P256P
     return CHIP_ERROR_NOT_IMPLEMENTED;
 }
 
+CHIP_ERROR ExtractAKIDFromX509Cert(const ByteSpan & certificate, MutableByteSpan & akid)
+{
+    return CHIP_ERROR_NOT_IMPLEMENTED;
```
Comment (bzbarsky-apple) [inline src/crypto/CHIPCryptoPALmbedTLS.cpp:1212]: Do we not need this implemented at some point?  This should at least be a TODO with a tracking issue filed.

File: src/crypto/tests/CHIPCryptoPALTest.cpp [hunk 24]
```diff
@@ -1795,8 +1795,232 @@ static void TestPubkey_x509Extraction(nlTestSuite * inSuite, void * inContext)
...
+        uint8_t certType = gTestCerts[i];
+
+        err = GetTestCert(certType, TestCertLoadFlags::kDERForm, cert);
+        NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+        err = GetTestCertAKID(certType, akidSpan);
+        NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+
+        err = ExtractAKIDFromX509Cert(cert, akidOut);
+        NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+        NL_TEST_ASSERT(inSuite, akidSpan.size() == akidOut.size() && memcmp(akidOut.data(), akidSpan.data(), akidSpan.size()) == 0);
```
Comment (bzbarsky-apple) [inline src/crypto/tests/CHIPCryptoPALTest.cpp:1821]:
```suggestion
        NL_TEST_ASSERT(inSuite, akidSpan.data_equal(akidOut));
```

File: src/crypto/tests/CHIPCryptoPALTest.cpp [hunk 25]
```diff
@@ -1795,8 +1795,232 @@ static void TestPubkey_x509Extraction(nlTestSuite * inSuite, void * inContext)
...
+    ByteSpan vendorReserved3Span;
+    ByteSpan vendorReserved4Span;
+    err = DeconstructAttestationElements(ByteSpan(attestationElementsTestVector), certificationDeclarationSpan,
+                                         attestationNonceSpan, timestampDeconstructed, firmwareInfoSpan, vendorReserved1Span,
+                                         vendorReserved2Span, vendorReserved3Span, vendorReserved4Span);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+
+    NL_TEST_ASSERT(inSuite, sizeof(certificationDeclaration) == certificationDeclarationSpan.size());
+    NL_TEST_ASSERT(inSuite,
+                   memcmp(certificationDeclarationSpan.data(), certificationDeclaration, sizeof(certificationDeclaration)) == 0);
```
Comment (bzbarsky-apple) [inline src/crypto/tests/CHIPCryptoPALTest.cpp:1950]: Might be simpler as:
```suggestion
    NL_TEST_ASSERT(inSuite, ByteSpan(certificationDeclaration).data_equal(certificationDeclarationSpan));
```
but either way, I guess.

File: src/include/platform/ConfigurationManager.h [hunk 26]
```diff
@@ -79,6 +79,13 @@ class ConfigurationManager
     CHIP_ERROR GetDeviceCertificate(uint8_t * buf, size_t bufSize, size_t & certLen);
     CHIP_ERROR GetDeviceIntermediateCACerts(uint8_t * buf, size_t bufSize, size_t & certsLen);
     CHIP_ERROR GetDevicePrivateKey(uint8_t * buf, size_t bufSize, size_t & keyLen);
+    CHIP_ERROR GetDeviceDACCertificate(uint8_t * buf, size_t bufSize, size_t & dacLen);
```
Comment (bzbarsky-apple) [inline src/include/platform/ConfigurationManager.h:82]: I know the existing APIs do this, but it might be better to have new additions look like:
```suggestion
    CHIP_ERROR GetDeviceDACCertificate(MutableByteSpan & buf);
```
Similar for other things here that have these sorts of 3-argument outparams.

File: src/include/platform/ConfigurationManager.h [hunk 27]
```diff
@@ -79,6 +79,13 @@ class ConfigurationManager
...
+    CHIP_ERROR GetDevicePAICertificate(uint8_t * buf, size_t bufSize, size_t & paiLen);
+    CHIP_ERROR GetDevicePAACertificate(const uint8_t * skid, size_t skidLen, uint8_t * buf, size_t bufSize, size_t & paaLen);
```
Comment (bzbarsky-apple) [inline src/include/platform/ConfigurationManager.h:84]: The (input only, right?) skid should _definitely_ be a `ByteSpan` here.  And this needs documentation explaining what that argument does.

File: src/include/platform/ConfigurationManager.h [hunk 28]
```diff
@@ -79,6 +79,13 @@ class ConfigurationManager
...
+    CHIP_ERROR GetDeviceCertificationDeclaration(uint8_t * buf, size_t bufSize, size_t & certDeclLen);
+    CHIP_ERROR GetDeviceAttestationSignature(const uint8_t * attestationElements, size_t attestationElementsSize,
+                                             const uint8_t * attestationChallenge, size_t attestationChallengeSize, uint8_t * buf,
+                                             size_t bufSize, size_t & signatureLen);
```
Comment (bzbarsky-apple) [inline src/include/platform/ConfigurationManager.h:88]: Again, use spans for input byte buffers.  And please document that this is implementing the part of the spec that signs the attestation message (NOT attestation elements!) and attestation challenge with the device attestation private key.  And that that last bit is why it's on the configuration manager.

File: src/lib/core/tests/TestCHIPTLV.cpp (hunk 19 above)
Comment (bzbarsky-apple) [reply src/lib/core/tests/TestCHIPTLV.cpp:2986]: Presumably because the old code is broken and asserts things about the contents of un-initialized memory....  But it should probably not happen in this PR.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/include/platform/internal/GenericConfigurationManagerImpl.h (hunk 21 above)
Comment (tcarmelveilleux) [reply src/include/platform/internal/GenericConfigurationManagerImpl.h:84]: @emargolis  Yes, just keep away from ConfigurationManager altogether.

File: src/crypto/CHIPCryptoPAL.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (mleisner) [reply src/crypto/CHIPCryptoPAL.cpp:831]: agreed, will do

File: src/credentials/tests/CHIPCert_test_vectors.cpp [hunk 29]
```diff
@@ -169,6 +169,43 @@ CHIP_ERROR GetTestCertPubkey(uint8_t certType, const uint8_t ** certPubkey, uint
...
+{
+    CHIP_ERROR err;
+
+#define SELECT_CERT(NAME)                                                                                                          \
+    do                                                                                                                             \
+    {                                                                                                                              \
+        if (certType == TestCert::k##NAME)                                                                                         \
+        {                                                                                                                          \
+            akid = ByteSpan(sTestCert_##NAME##_AuthorityKeyId, sTestCert_##NAME##_AuthorityKeyId_Len);                             \
+            ExitNow(err = CHIP_NO_ERROR);                                                                                          \
```
[the remark this answers is not in the document]
Comment (mleisner) [reply src/credentials/tests/CHIPCert_test_vectors.cpp:182]: Agreed, also remove err, and just return error code (I personally like this better).

File: src/crypto/CHIPCryptoPAL.cpp (hunk 3 above)
Comment (mleisner) [reply src/crypto/CHIPCryptoPAL.cpp:748]: agreed

File: src/crypto/CHIPCryptoPAL.cpp [hunk 30]
```diff
@@ -648,5 +649,185 @@ CHIP_ERROR GenerateCompressedFabricId(const Crypto::P256PublicKey & root_public_
...
+    ReturnErrorOnFailure(hashStream.Finish(messageDigestSpan));
+
+    ReturnErrorOnFailure(ECDSA_sign_hash(messageDigestSpan.data(), messageDigestSpan.size(), out_signature));
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR P256PublicKey::ECDSA_validate_attestation_data(const ByteSpan & attestationElements,
+                                                          const ByteSpan & attestationChallenge,
+                                                          const P256ECDSASignature & out_signature) const
```
[the remark this answers is not in the document]
Comment (mleisner) [reply src/crypto/CHIPCryptoPAL.cpp:671]: agreed

**Comment by tcarmelveilleux** [comment]
Note stale, still in progress from @mleisner

**Comment by tcarmelveilleux** [comment]
From offline convos w/ @mleisner, updates to come in next few days. Action occurring in fork daily, but not ready for mainline. Not stale.
